<a href="https://colab.research.google.com/github/PineappleBingo/CSC448-AI/blob/main/CSC448_Seo_Jinho_AS03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [75]:
# Imports section
import io
import matplotlib.pyplot as plt
import pandas as pd
import requests
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import train_test_split

## Part 1. Loading the dataset

In [76]:
# Using pandas load the dataset (load remotely, not locally)
CSV_URL = "https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv"
# s = requests.get(CSV_URL).content
# df = pd.read_csv(io.StringIO(s.decode('utf-8')))
df = pd.read_csv(CSV_URL)

# Output the first 15 rows of the data
print(df.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print(df.info())
# Shape
print("\nRow x Columns:", df.shape)
# column names
print("\nColumn Names:", df.columns)


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [77]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = df[["Temperature °C", "Mols KCL"]]
Y = df["Size nm^3"]

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=10)
# Set random_state(a.k.a seed) so that we can reproduce our results

print("X_train Shape:", X_train.shape)
print("X_test Shape:", X_test.shape)
print("Y_train Shape:", Y_train.shape)
print("Y_test Shape:", Y_test.shape)


X_train Shape: (900, 2)
X_test Shape: (100, 2)
Y_train Shape: (900,)
Y_test Shape: (100,)


## Part 3. Perform a Linear Regression

In [78]:
# Use sklearn to train a model on the training set

clf = LinearRegression()
clf.fit(X_train, Y_train)

# # Make predictions using the testing set
# y_pred = clf.predict(X_test)

# # The mean squared error
# print("Mean squared error: %.2f" % mean_squared_error(Y_test, y_pred))
# # The coefficient of determination: 1 is perfect prediction
# print("Coefficient of determination: %.2f" % r2_score(Y_test, y_pred))

# # Plot outputs
# plt.scatter(X_test, Y_test, color="black")
# plt.plot(X_test, y_pred, color="blue", linewidth=3)

# plt.xticks(())
# plt.yticks(())

# plt.show()
# 
# https://scikit-learn.org/stable/auto_examples/linear_model/plot_ols.html#sphx-glr-auto-examples-linear-model-plot-ols-py
# -----------------------------------------------

# Create a sample datapoint and predict the output of that sample with the trained model
sample = [[500, 500], [600, 600], [700, 700], [800, 800]]
preds_y = clf.predict(sample)

for idx, pred in enumerate(preds_y):
    print(str(sample[idx]) + " -> Preds_Y:", pred)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
print("\nScore:", clf.score(X_test, Y_test))

# Explanation
# Using linearRegression algorithm, we traind the data set X and Y, predicting new Ys depends on sample datapoints.
# Based on the scores, we can tell how accurately the chosen algorithm woulde predict outputs by given trained dataset
# ; the hight score means higher accuratcy in predecting outouts. In our case, score is 88% which means linearRegression 
# algorithm has 88% accuracy to predict expanding size of the slime by temperature and amounts of KCL after one day.
   
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
print("\nCoefficients: \n", clf.coef_[0])
print("Intercept:", clf.intercept_ )


[500, 500] -> Preds_Y: 534158.5946029222
[600, 600] -> Preds_Y: 725035.7824817223
[700, 700] -> Preds_Y: 915912.9703605224
[800, 800] -> Preds_Y: 1106790.1582393225

Score: 0.8838673863856773

Coefficients: 
 876.5603799396862
Intercept: -420227.3447910782


D:\Program Files\Python310\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [ ]:
# H(x)

\begin{gather*}
h(x) = 876.56x -420227.34
\end{gather*}

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [79]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance

## Part 5. Using Polynomial Regression

In [80]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.